[quant][core][gpu][bug-fix] Added additional caching support in quantized cudnn add_relu op

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75772

Summary:
Previous caching strategy for quantized cudnn add_relu operator was insufficient
as it did not properly record all the necesary information. This PR adds several
items to the CacheKey (e.g., input sizes, input dimensions, etc..) that enables
proper caching

Test plan:
```
python test/test_quantization.py -k test_qadd_relu_cudnn
```

Differential Revision: [D35634278](https://our.internmc.facebook.com/intern/diff/D35634278)

dzdang committed:
[quant][core][gpu][bug-fix] Added additional caching support in quauntized cudnn add_relu op

Summary:
Previous caching strategy for quantized cudnn add_relu operator was insufficient
as it did not properly record all the necesary information. This PR adds several
items to the CacheKey (e.g., input sizes, input dimensions, etc..) that enables
proper caching

Test plan:
```
python test/test_quantization.py -k test_qadd_relu_cudnn
```

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][bug-fix] Added additional caching support in quantized cudnn add_relu op"


Summary:
Previous caching strategy for quantized cudnn add_relu operator was insufficient
as it did not properly record all the necesary information. This PR adds several
items to the CacheKey (e.g., input sizes, input dimensions, etc..) that enables
proper caching

Test plan:
```
python test/test_quantization.py -k test_qadd_relu_cudnn
```

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35634278).

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:
```diff
@@ -21,16 +22,43 @@
+    bool deterministic, bool allow_tf32) {
+  // operator datatype needs to be int32 for int8 matmul, but we can
+  // set the datatype for output tensor to int32 or fp32
+  memset(params, 0, sizeof(AddParams));
```
do we need to do this?

dzdang replied:
I copied that over from https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/ConvShared.cpp#L89. 
I think it's used for initialization and to set the remaining values of `input_*_size` to zero. Without it, I think we'd just need an additional loop to initialize the remaining values.

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:
```diff
@@ -21,16 +22,43 @@
+void setAddParams(
+    AddParams* params, const at::Tensor& input_a, const at::Tensor& input_b,
+    bool deterministic, bool allow_tf32) {
+  // operator datatype needs to be int32 for int8 matmul, but we can
+  // set the datatype for output tensor to int32 or fp32
```
looks like this comment is not related?

dzdang committed:
Update on "[quant][core][gpu][bug-fix] Added additional caching support in quantized cudnn add_relu op"


Summary:
Previous caching strategy for quantized cudnn add_relu operator was insufficient
as it did not properly record all the necesary information. This PR adds several
items to the CacheKey (e.g., input sizes, input dimensions, etc..) that enables
proper caching

Test plan:
```
python test/test_quantization.py -k test_qadd_relu_cudnn
```

Differential Revision: [D35634278](https://our.internmc.facebook.com/intern/diff/D35634278)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35634278).
